rust/candid_parser/src/bindings/javascript.rs

```diff
@@ -237,38 +245,84 @@ fn pp_actor<'a>(ty: &'a Type, recs: &'a BTreeSet<&'a str>) -> RcDoc<'a> {
     }
 }
 
+fn pp_deprecation_comment<'a>() -> RcDoc<'a> {
+    str("/**").append(RcDoc::hardline())
+        .append(" * @deprecated Import IDL types directly from this module instead of using this factory function.")
+        .append(RcDoc::hardline())
+        .append(" */")
+        .append(RcDoc::hardline())
+}
+
 pub fn compile(env: &TypeEnv, actor: &Option<Type>) -> String {
     match actor {
         None => {
             let def_list: Vec<_> = env.0.iter().map(|pair| pair.0.as_ref()).collect();
             let recs = infer_rec(env, &def_list).unwrap();
-            let doc = pp_defs(env, &def_list, &recs);
-            doc.pretty(LINE_WIDTH).to_string()
+            let import_doc = str("import { IDL } from '@dfinity/candid';");
+            let doc = pp_defs(env, &def_list, &recs, true);
+            let result = import_doc
+                .append(RcDoc::hardline())
+                .append(RcDoc::hardline())
+                .append(doc)
+                .pretty(LINE_WIDTH)
+                .to_string();
+
+            result
         }
         Some(actor) => {
             let def_list = chase_actor(env, actor).unwrap();
             let recs = infer_rec(env, &def_list).unwrap();
-            let defs = pp_defs(env, &def_list, &recs);
             let init = if let TypeInner::Class(ref args, _) = actor.as_ref() {
                 args.as_slice()
             } else {
                 &[][..]
             };
-            let actor = kwd("return").append(pp_actor(actor, &recs)).append(";");
-            let body = defs.append(actor);
-            let doc = str("export const idlFactory = ({ IDL }) => ")
-                .append(enclose_space("{", body, "};"));
-            // export init args
+
+            let import_doc = str("import { IDL } from '@dfinity/candid';");
+            let defs = pp_defs(env, &def_list, &recs, true);
+            let actor = pp_actor(actor, &recs);
+
+            let idl_service = str("export const idlService = ")
+                .append(actor.clone())
+                .append(";");
+
+            let idl_init_args = str("export const idlInitArgs = ")
+                .append(pp_rets(init))
+                .append(";");
+
+            let idl_factory_return = kwd("return").append(actor).append(";");
+            let idl_factory_body = pp_defs(env, &def_list, &recs, false).append(idl_factory_return);
+            let idl_factory_doc = pp_deprecation_comment()
+                .append(str("export const idlFactory = ({ IDL }) => "))
```

Comment:
I would now rename the `IDL` parameter to something like:

```suggestion
                .append(str("export const idlFactory = ({ IDL: IDLModule }) => "))
```

to avoid variable shadowing and make sure we are using the IDL passed as argument inside the `idlFactory` and the `init` functions, and **not** the top level imported `IDL` obect instead.

This requires us to pass a new argument (e.g. `idl_module_name: 'a str`) around all the nested functions until `pp_ty`, but I think it's worth it. Maybe you can find a better way to avoid passing this argument around.

Reply:
I think this is fine as-is to avoid the complexity.

Avoiding variable shadowing is a stylistic choice and this function is deprecated now anyway. I don't think it's worth introducing additional complexity for a stylistic choice in a deprecated function.